README.md
bf42
====

Learning Love2D using [@josefnpat's](https://github.com/josefnpat) excellent [Love2D tutorial](http://cupm.net/public/love2dtutorial/)

## Running the game

Grab the [assets (7z)](http://cupm.net/public/love2dtutorial/1942gb_v1.0.7z), unzip `assets.zip` to the cloned folder, then:

    love main.lua


CHANGELOG.txt
This changelog documents changes from the original tutorial.

- Added a gameover screen
- Added a boss (just a triple enemy).
- Added ability for enemies (including bosses) to shoot bullets.
- Added a powerup shield which adds a shield for the player, and protects player from one collision with a bullet or enemy.
- Added a triple bullet powerup(tbp) - player shoots 3 bullets


game.lua
game = {}

function game.load()
  -- background init
  game.clock = 0
  game.gameover_time = 0
  game.boss_dt = 0
  game.boss_bullets = {}
  -- enemy init
  game.enemy_size = imgs["enemy"]:getWidth()
  game.enemies = {}
  game.enemy_dt = 0
  game.enemy_rate = 2
  game.enemy_bullets = {}
  -- boss
  game.num_bosses = 5
  game.bosses = {}
  -- player init
  game.player_size = imgs["player"]:getWidth()
  game.playerx = (160/2)*scale
  game.playery = (144-12)*scale
  -- bullet init
  game.ammo = 10
  game.recharge_dt = 0
  game.recharge_rate = 1
  game.bullet_size = imgs["bullet"]:getWidth()
  game.bullets = {}
  -- shields init
  game.num_shields = 2
  game.shield_dt = 0
  game.shields = {}
  game.player_has_shield = false
  -- triple bullet powerup (tbp)
  game.num_tbps = 2
  game.tbp_dt = 0
  game.tbps = {}
  game.player_has_tbp = false
  -- info init
  game.score = 0
end

function game.draw()
  -- Draw moving background
  for i = 0,4 do
    for j = -1,4 do
      love.graphics.draw(imgs["background"],
                           i*32*scale,
                           (j+game.clock%1)*32*scale,
                           0,scale,scale)
    end
  end
  -- Draw enemies
  for _,v in ipairs(game.enemies) do
    love.graphics.draw(imgs["enemy"],
                         v.x,v.y,
                         0,scale,scale,
                         game.enemy_size/2,game.enemy_size/2)
    if debug then love.graphics.circle("line",v.x,v.y,(game.enemy_size/2)*scale) end
  end

  -- Draw bosses (just a triple enemy)
  for i = -1,1 do
    for _,v in ipairs(game.bosses) do
      love.graphics.draw(imgs["enemy"],
                           v.x+(game.enemy_size*i),v.y,
                           0,scale,scale,
                           game.enemy_size/2,game.enemy_size/2)
    if debug then love.graphics.circle("line",v.x,v.y,game.enemy_size/2*scale) end
    end
  end
  -- Draw player
  love.graphics.draw(imgs["player"],
                       game.playerx,game.playery,
                       0,scale,scale,
                       game.player_size/2,game.player_size/2)
  if debug then love.graphics.circle("line",game.playerx,game.playery,game.player_size/2*scale) end
  if game.player_has_shield then
    love.graphics.setColor(255,0,0)
    for i=0,1 do
      love.graphics.circle("line",game.playerx,game.playery,(game.player_size/2 + i)*scale)
    end
    love.graphics.setColor(255,255,255)
  end
  -- Draw game.bullets
  for _,v in ipairs(game.bullets) do
    love.graphics.draw(imgs["bullet"],
                         v.x,v.y,
                         0,scale,scale,
                         game.bullet_size/2,game.bullet_size/2)
    if debug then love.graphics.circle("line",v.x,v.y,game.bullet_size/2*scale) end
  end

  -- Draw enemy bullets
  for _,v in ipairs(game.enemy_bullets) do
    love.graphics.draw(imgs["bullet"],
                         v.x,v.y,
                         0,scale,scale,
                         game.bullet_size/2,game.bullet_size/2)
    if debug then love.graphics.circle("line",v.x,v.y,game.bullet_size/2*scale) end
  end

  -- Draw boss bullets
  for _,v in ipairs(game.boss_bullets) do
    love.graphics.draw(imgs["bullet"],
                         v.x,v.y,
                         0,scale,scale,
                         game.bullet_size/2,game.bullet_size/2)
    if debug then love.graphics.circle("line",v.x,v.y,game.bullet_size/2*scale) end
  end

  -- Draw shields
  for _,v in ipairs(game.shields) do
    love.graphics.setColor(0,0,0)
    love.graphics.circle("fill",v.x,v.y,game.bullet_size/2*scale)
    love.graphics.setColor(255,255,255)
  end

  -- Draw tbps
  for _,v in ipairs(game.tbps) do
    love.graphics.setColor(255,255,255)
    love.graphics.circle("fill",v.x,v.y,game.bullet_size/2*scale)
    love.graphics.setColor(255,255,255)
  end
  
  -- Draw game info
  love.graphics.setColor(fontcolor.r,fontcolor.g,fontcolor.b)
  love.graphics.printf(
    "score:"..game.score..
    " ammo:"..game.ammo,
  0,0,love.graphics.getWidth(),"center")

  if debug then love.graphics.print(
    "enemies: "..#game.enemies..
    "\nbullets:"..#game.bullets..
    "\nenemy_rate:"..game.enemy_rate..
    "\nFPS:"..love.timer.getFPS(),
  0,14*scale) end
  love.graphics.setColor(255,255,255)
end

-- Distance formula.
function game.dist(x1,y1,x2,y2)
  return math.sqrt( (x1 - x2)^2 + (y1 - y2)^2 )
end

function game.update(dt)
  -- clock for background
  game.clock = game.clock + dt
  -- Update game.enemies
  game.enemy_dt = game.enemy_dt + dt
  -- Update game.bosses
  game.boss_dt = game.boss_dt + dt
  -- Update shields
  game.shield_dt = game.shield_dt + dt
  -- Update tbps
  game.tbp_dt = game.tbp_dt + dt
  
  -- Enemy spawn
  if game.enemy_dt > game.enemy_rate then
    game.enemy_dt = game.enemy_dt - game.enemy_rate
    game.enemy_rate = game.enemy_rate - 0.01 * game.enemy_rate
    local enemy = {}
    enemy.x = math.random((8)*scale,(160-8)*scale)
    enemy.y = -game.enemy_size
    enemy.num_bullets = 1
    table.insert(game.enemies,enemy)
  end

  -- Boss spawn
  if game.boss_dt > 5 then
    local boss = {}
    boss.x = math.random((24)*scale,(160-24)*scale)
    boss.y = -game.enemy_size
    boss.num_bullets = 1
    table.insert(game.bosses,boss)
    game.num_bosses = game.num_bosses - 1
    game.boss_dt = 0
  end
  
  -- Update enemy
  for ei,ev in ipairs(game.enemies) do
    ev.y = ev.y + 70*dt*scale
    if ev.y > 20*scale and ev.num_bullets > 0  then
      local ebullet = {}
      ebullet.x = ev.x
      ebullet.y = ev.y
      table.insert(game.enemy_bullets,ebullet)
      ev.num_bullets = ev.num_bullets - 1
    end 
    if ev.y > 144*scale then
      table.remove(game.enemies,ei)
    end
    -- If a player gets too close to enemy
    if game.dist(game.playerx,game.playery,ev.x,ev.y) < (12+8)*scale then
      if game.player_has_shield then
        game.player_has_shield = false
        table.remove(game.enemies,ei)
      else
        game.gameover_time = game.clock
        gameover.load()
        state = "gameover"
      end
    end
  end

  -- Update boss
  for ei,ev in ipairs(game.bosses) do
    ev.y = ev.y + 70*dt*scale
    if ev.y > 20*scale and ev.num_bullets > 0  then
      for i=-1,1 do
        local bbullet = {}
        bbullet.x = ev.x+(i*game.enemy_size)
        bbullet.y = ev.y
        table.insert(game.boss_bullets,bbullet)
      end      
      ev.num_bullets = ev.num_bullets - 1
    end 
    if ev.y > 144*scale then
      table.remove(game.bosses,ei)
    end
    -- If a player gets too close to boss
    if game.dist(game.playerx,game.playery,ev.x,ev.y) < (12+8)*scale then
      if game.player_has_shield then
        game.player_has_shield = false
        table.remove(game.bosses,ei)
      else
        game.gameover_time = game.clock

        gameover.load()
        state = "gameover"
      end
    end
  end

  -- Spawn shields
  if game.num_shields > 0 and game.shield_dt > 10 and not game.player_has_shield then
    local shield = {}
    shield.x = math.random((24)*scale,(160-24)*scale)
    shield.y = -game.enemy_size
    table.insert(game.shields,shield)
    game.num_shields = game.num_shields - 1
    game.shield_dt = 0
  end

  -- Update shields
  for ei,ev in ipairs(game.shields) do
    ev.y = ev.y + 70*dt*scale
    if ev.y > 144*scale then
      table.remove(game.shields,ei)
    end
    -- If a player gets too close to shield
    if game.dist(game.playerx,game.playery,ev.x,ev.y) < (12+8)*scale then
      game.player_has_shield = true
      table.remove(game.shields,ei)
    end
  end  

  -- Spawn tbps
  if game.num_tbps > 0 and game.tbp_dt > 15 and not game.player_has_tbp then
    local tbp = {}
    tbp.x = math.random((24)*scale,(160-24)*scale)
    tbp.y = -game.enemy_size
    table.insert(game.tbps,tbp)
    game.num_tbps = game.num_tbps - 1
    game.tbp_dt = 0
  end

  -- Update tbps
  for ei,ev in ipairs(game.tbps) do
    ev.y = ev.y + 70*dt*scale
    if ev.y > 144*scale then
      table.remove(game.tbps,ei)
    end
    -- If a player gets too close to tbp
    if game.dist(game.playerx,game.playery,ev.x,ev.y) < (12+8)*scale then
      game.player_has_tbp = true
      table.remove(game.tbps,ei)
    end
  end  
  
  -- Update player movement
  if love.keyboard.isDown("right") then
    game.playerx = game.playerx + 100*dt*scale
  end
  if love.keyboard.isDown("left") then
    game.playerx = game.playerx - 100*dt*scale
  end
  if love.keyboard.isDown("up") then
    game.playery = game.playery - 100*dt*scale
  end
  if love.keyboard.isDown("down") then
    game.playery = game.playery + 100*dt*scale
  end
  -- Keep the player on the map
  if game.playerx > 160*scale then
    game.playerx = 160*scale
  end
  if game.playerx < 0 then
    game.playerx = 0
  end
  if game.playery < 0 then
    game.playery = 0
  end
  if game.playery > 144*scale then
    game.playery = 144*scale
  end
  
  -- Update bullets
  for bi,bv in ipairs(game.bullets) do
    if game.player_has_tbp then
      bv.x = bv.x + 40*dt*bv.tbp_id
    end
    bv.y = bv.y - 100*dt*scale
    if bv.y < 0 then
      table.remove(game.bullets,bi)
    end
    -- Update bullets with game.enemies
    for ei,ev in ipairs(game.enemies) do
      if game.dist(bv.x,bv.y,ev.x,ev.y) < (2+8)*scale then
        game.score = game.score + 1
        table.remove(game.enemies,ei)
        table.remove(game.bullets,bi)
      end
    end
    -- Update bullets with game.bosses
    for ei,ev in ipairs(game.bosses) do
      if game.dist(bv.x,bv.y,ev.x,ev.y) < (2+8)*scale then
        game.score = game.score + 3 -- extra points for boss
        table.remove(game.bosses,ei)
        table.remove(game.bullets,bi)
      end
    end
  end

  -- Update enemy bullets
  for bi,bv in ipairs(game.enemy_bullets) do
    bv.y = bv.y + 100*dt*scale
    if bv.y > 144*scale then
      table.remove(game.enemy_bullets,bi)
    end
    -- see if enemy bullet hit player
    if game.dist(game.playerx,game.playery,bv.x,bv.y) < (12+8)*scale then
      if game.player_has_shield then
        game.player_has_shield = false
        table.remove(game.enemy_bullets,bi)
      else
        game.gameover_time = game.clock
        gameover.load()
        state = "gameover"
      end
    end
  end

  -- Update boss bullets
  for bi,bv in ipairs(game.boss_bullets) do
    bv.y = bv.y + 120*dt*scale  -- faster bullets for bosses
    if bv.y > 144*scale then
      table.remove(game.boss_bullets,bi)
    end
    -- see if boss bullet hit player
    if game.dist(game.playerx,game.playery,bv.x,bv.y) < (12)*scale then
      if game.player_has_shield then
        game.player_has_shield = false
        table.remove(game.boss_bullets,bi)
      else
        game.gameover_time = game.clock
        gameover.load()
        state = "gameover"
      end
    end
  end
  
  -- Update player ammunition
  game.recharge_dt = game.recharge_dt + dt
  if game.recharge_dt > game.recharge_rate then
    game.recharge_dt = game.recharge_dt - game.recharge_rate
    game.ammo = game.ammo + 1
    if game.ammo > 10 then
      game.ammo = 10
    end
  end
end

function game.keypressed(key)
  -- Shoot a bullet
  if key == " " and game.ammo > 0 then
    love.audio.play(shoot)
    game.ammo = game.ammo - 1
    if game.player_has_tbp then
      for i=-1,1 do
        local bullet = {}
        bullet.x = game.playerx
        bullet.y = game.playery
        bullet.tbp_id = i
        table.insert(game.bullets,bullet)
      end
    else
      local bullet = {}
      bullet.x = game.playerx
      bullet.y = game.playery
      table.insert(game.bullets,bullet)
    end
  end
end


splash.lua
splash = {}

function splash.load()
  splash.dt_temp = 0
end

function splash.draw()
  love.graphics.draw(imgs["title"],0,(splash.dt_temp-1)*32*scale,0,scale,scale)
  love.graphics.setColor(fontcolor.r,fontcolor.g,fontcolor.b)
  -- Show after 2.5 seconds
  if splash.dt_temp == 2.5 then
    love.graphics.printf("Press Start",
      0,80*scale,love.graphics.getWidth(),"center")
  end
  -- If previous game
  if game.score ~= 0 then
    love.graphics.printf("Score:"..game.score,0,96*scale,160*scale,"center")
  end
  -- Reset the color
  love.graphics.setColor(255,255,255)
end

function splash.update(dt)
  -- Update dt_temp
  splash.dt_temp = splash.dt_temp + dt
  -- Wait 2.5 seconds, then stop in place.
  if splash.dt_temp > 2.5 then
    splash.dt_temp = 2.5
  end
end

function splash.keypressed(key)
  -- Change to game state, and init game.
  state = "game"
  game.load()
end


gameover.lua
gameover = {}

function gameover.load()
  gameover.dt_temp = 0
end

function gameover.draw()
  -- love.graphics.draw(imgs["title"],0,(gameover.dt_temp-1)*32*scale,0,scale,scale)
  love.graphics.setColor(fontcolor.r,fontcolor.g,fontcolor.b)
  -- Show after 2.5 seconds
  love.graphics.printf("Gameover",
    0,(gameover.dt_temp-1)*32*scale,love.graphics.getWidth(),"center")
  -- If previous game
  if game.score ~= 0 then
    love.graphics.printf("Score:"..game.score,0,96*scale,160*scale,"center")
  end
  -- Reset the color
  love.graphics.setColor(255,255,255)
end

function gameover.update(dt)
  -- Update dt_temp
  gameover.dt_temp = gameover.dt_temp + dt
  if debug then love.graphics.printf(gameover.dt_temp,0,80*scale,80*scale,"center") end
  -- Wait 2.5 seconds, then stop in place.
  if gameover.dt_temp > 2.5 then
    gameover.dt_temp = 2.5
  end
end

function gameover.keypressed(key)
  -- Change to game state, and init game.
  state = "splash"
  splash.load()
end


main.lua
debug = false

require('splash')
require('game')
require ('gameover')

function love.load()
  -- Load images (global assets)
  img_fn = {"bullet","enemy","player","title","background"}
  imgs = {}
  for _,v in ipairs(img_fn) do
    imgs[v]=love.graphics.newImage("assets/"..v..".gif")
  end
  
  -- Set filter to nearest
  for _,v in pairs(imgs) do
    v:setFilter("nearest","nearest")
  end

  -- Play music and loop it.
  music = love.audio.newSource( "assets/music.ogg" , "stream" )
  music:setLooping(true)
  love.audio.play(music)

  -- load shoot sound
  shoot = love.audio.newSource( "assets/shoot.ogg" , "static" )

  -- Initialize font, and set it.
  font = love.graphics.newFont("assets/font.ttf",14*scale)
  love.graphics.setFont(font)

  -- define colors (global assets)
  bgcolor = {r=148,g=191,b=19}
  fontcolor = {r=46,g=115,b=46}

  -- initial state
  state = "splash"  
  
  --load the splash
  splash.load()
  game.load()
end

function love.draw()
  -- Set color
  love.graphics.setColor(bgcolor.r,bgcolor.g,bgcolor.b)
  -- Draw rectangle for background
  love.graphics.rectangle("fill",
    0,0 ,love.graphics.getWidth(),love.graphics.getHeight())
  -- Return the color back to normal.
  love.graphics.setColor(255,255,255)
  -- Call the state's draw function
  if state == "splash" then
    splash.draw()
  elseif state == "game" then
    game.draw()
  elseif state == "gameover" then
    gameover.draw()
  end
end

function love.update(dt)
  -- Call the state's update function
  if state == "splash" then
    splash.update(dt)
  elseif state == "game" then
    game.update(dt)
  elseif state == "gameover" then
    gameover.update(dt)
  end

end

function love.keypressed(key)
  -- Call the state's keypressed function
  if state == "splash" then
    splash.keypressed(key)
  elseif state == "game" then
    game.keypressed(key)
  elseif state == "gameover" then
    gameover.keypressed(key)
  end
  if key == "`" then
    debug = not debug
  end
end
